op-supervisor/supervisor/backend/source/chain.go

```diff
@@ -0,0 +1,95 @@
+package source
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-service/client"
+	"github.com/ethereum-optimism/optimism/op-service/dial"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/sources"
+	"github.com/ethereum-optimism/optimism/op-service/sources/caching"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+// TODO(optimism#10999) Make these configurable and a sensible default
+const epochPollInterval = 30 * time.Second
+const pollInterval = 2 * time.Second
+const trustRpc = false
+const rpcKind = sources.RPCKindStandard
+
+type Metrics interface {
+	CacheAdd(chainID *big.Int, label string, cacheSize int, evicted bool)
+	CacheGet(chainID *big.Int, label string, hit bool)
+}
+
+// ChainMonitor monitors a source L2 chain, retrieving the data required to populate the database and perform
+// interop consolidation. It detects and notifies when reorgs occur.
+type ChainMonitor struct {
+	headMonitor *HeadMonitor
+}
+
+func NewChainMonitor(ctx context.Context, logger log.Logger, genericMetrics Metrics, rpc string) (*ChainMonitor, error) {
+	// First dial a simple client and get the chain ID so we have a simple identifier for the chain.
+	ethClient, err := dial.DialEthClientWithTimeout(ctx, 10*time.Second, logger, rpc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to rpc %v: %w", rpc, err)
+	}
+	chainID, err := ethClient.ChainID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load chain id for rpc %v: %w", rpc, err)
+	}
+	logger = logger.New("chainID", chainID)
+	m := newChainMetrics(chainID, genericMetrics)
+	cl, err := newClient(ctx, logger, m, rpc, ethClient.Client(), pollInterval, trustRpc, rpcKind)
+	if err != nil {
+		return nil, err
+	}
+	logger.Info("Monitoring chain", "rpc", rpc)
```

Comment:
This may have a secret in it

Reply:
Remove the RPC url from this, though we log the rpc URL all over the place - it's in most of the errors. (eg https://github.com/ethereum-optimism/optimism/blob/61b2b36cc1d317b72a43ed1832f5603c3527be95/op-service/dial/dial.go#L70).  Knowing which RPC is failing is going to be really important here too so that will be hard to avoid. We likely would need to sanitise the URL in some way.